src/actions/eval-queue-saga.js

```diff
@@ -0,0 +1,144 @@
+import CodeMirror from "codemirror";
+import {
+  take,
+  actionChannel,
+  put,
+  call,
+  select,
+  flush
+} from "redux-saga/effects";
+
+import {
+  NONCODE_EVAL_TYPES,
+  RUNNABLE_CHUNK_TYPES
+} from "../state-schemas/state-schema";
+
+import { setKernelState } from "./eval-actions";
+
+import {
+  loadingLanguageConsoleMsg,
+  addInputToConsole,
+  evalTypeConsoleError,
+  pluginParseError
+} from "./console-message-actions";
+
+import messagePasserEditor from "../redux-to-port-message-passer";
+import generateRandomId from "../tools/generate-random-id";
+
+// helpers
+
+export const languageReady = (state, lang) =>
+  Object.keys(state.loadedLanguages).includes(lang);
+
+export const languageKnown = (state, lang) =>
+  Object.keys(state.languageDefinitions).includes(lang);
+
+const definedEvalType = (state, lang) =>
+  languageReady(state, lang) ||
+  languageKnown(state, lang) ||
+  RUNNABLE_CHUNK_TYPES.includes(lang) ||
+  NONCODE_EVAL_TYPES.includes(lang);
+
+export const languageNeedsLoading = (state, lang) =>
+  languageKnown(state, lang) && !languageReady(state, lang);
+
+// sender
+
+export function sendTaskToEvalFrame(taskType, payload) {
+  const taskId = generateRandomId();
+  messagePasserEditor.postMessage(
+    taskType,
+    Object.assign({}, payload, { taskId })
+  );
+  return taskId;
+}
+
+// sagas
+export function* triggerEvalFrameTask(taskType, payload) {
+  const taskId = yield call(sendTaskToEvalFrame, taskType, payload);
```

Comment:
it keeps the saga pure which aids with testing. one of the big ideas of redux-saga that i like is that rather than directly calling functions, you just declare the thing you want to happen. by doing this, you don't need to mock these functions in testing, you just say "function F should have been called with args X,Y,Z". And using the "provide" helper, you can say that during testing a declared call of F with XYZ should yield P.

so pure! :-)

Reply:
sure, that makes sense